### PR TITLE
A serializer class that is pulled by the client, instead of pushing to the client

### PR DESCRIPTION
**Pull request by pleroy**

### CODE REVIEW

File: base/pull_serializer.hpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#include "google/protobuf/message.h"
+#include "google/protobuf/io/zero_copy_stream.h"
+
+namespace principia {
+namespace base {
+
+// An output stream based on two arrays that delegates to a function the
+// handling of the case where one array is full.  It calls the |on_full|
+// function passed at construction and proceeds with filling the other array.
+class DelegatingTwoArrayOutputStream
```
Comment (eggrobin) [inline base/pull_serializer.hpp]: Probably internal until some convincing general use case appears

File: base/pull_serializer.hpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  int position_;
+  int last_returned_size_;   // How many bytes we returned last time Next()
+                             // was called (used for error checking only).
+};
+
+// This class support serialization which is "pulled" by the client.  That is,
+// the client creates a |PullSerializer| object, calls |Start| to start the
+// serialization process, repeatedly calls |Get| to obtain a chunk of data, and
+// finally destroys the |PullSerializer|.
+class PullSerializer {
```
Comment (eggrobin) [inline base/pull_serializer.hpp]: Comment: for use in memory-critical contexts

File: base/pull_serializer.hpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  // Similar to a StringPiece.  |data| is not owned.
+  struct Data {
+    Data(not_null<std::uint8_t const*> const data, int const size);
+    not_null<std::uint8_t const*> const data;
+    int const size;
+  };
+
+  // The |size| of the data objects returned by |Get| are never greater than
+  // |max_size|.
+  explicit PullSerializer(int const max_size);
```
Comment (eggrobin) [inline base/pull_serializer.hpp]: Comment on the memory footprint

File: base/pull_serializer.hpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+// serialization process, repeatedly calls |Get| to obtain a chunk of data, and
+// finally destroys the |PullSerializer|.
+class PullSerializer {
+ public:
+  // Similar to a StringPiece.  |data| is not owned.
+  struct Data {
+    Data(not_null<std::uint8_t const*> const data, int const size);
+    not_null<std::uint8_t const*> const data;
+    int const size;
+  };
```
Comment (eggrobin) [inline base/pull_serializer.hpp]: Directly under `base`, find a nice name, ideally containing "byte".
`ByteStrand`?

Comment (eggrobin) [inline base/pull_serializer.hpp]: Use in the hex encode/decode methods or make an issue.

File: base/pull_serializer.hpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+namespace principia {
+namespace base {
+
+// An output stream based on two arrays that delegates to a function the
+// handling of the case where one array is full.  It calls the |on_full|
+// function passed at construction and proceeds with filling the other array.
+class DelegatingTwoArrayOutputStream
+    : public google::protobuf::io::ZeroCopyOutputStream {
+ public:
+  // The stream is supported by |data| which has size |size|.  Internally,
```
Comment (eggrobin) [inline base/pull_serializer.hpp]: The code seems to be consistent with `data` having size `size << 1`.

File: base/pull_serializer.hpp (hunk 4 above)
Comment (eggrobin) [inline base/pull_serializer.hpp]: `Bytes` seems nice.

File: base/pull_serializer.hpp [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+  // pieces is full.  Output can continue with the other piece.  |on_full| is
+  // expected to somehow consume the data of the array.  Note that two
+  // successive calls to |on_full| are always passed different pointers, so the
+  // caller can assume that it's fine to operate on that pointer in between two
+  // calls to |on_full|.
+  DelegatingTwoArrayOutputStream(
+      not_null<std::uint8_t*> data,
+      int const size,
+      std::function<void(not_null<std::uint8_t const*> const data,
+                         int const size)> on_full);
```
Comment (eggrobin) [inline base/pull_serializer.hpp]: `data, size` should be a `Bytes`.
`on_full` should take `Bytes`.

File: base/pull_serializer.hpp (hunk 1 above)
Comment (pleroy) [inline base/pull_serializer.hpp]: Done.

File: base/pull_serializer.hpp (hunk 3 above)
Comment (pleroy) [inline base/pull_serializer.hpp]: Done

File: base/pull_serializer.hpp (hunk 2 above)
Comment (pleroy) [inline base/pull_serializer.hpp]: Done.

File: base/pull_serializer.hpp (hunk 4 above)
Comment (pleroy) [inline base/pull_serializer.hpp]: Added `Bytes`.

File: base/pull_serializer.hpp (hunk 5 above)
Comment (pleroy) [inline base/pull_serializer.hpp]: Done.

File: base/pull_serializer.hpp (hunk 6 above)
Comment (pleroy) [inline base/pull_serializer.hpp]: Done.

File: base/pull_serializer_body.hpp [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+PullSerializer::~PullSerializer() {
+  if (thread_ != nullptr) {
+    thread_->join();
+  }
+}
+
+void PullSerializer::Start(
+    base::not_null<google::protobuf::Message const*> const message) {
+  CHECK(thread_ == nullptr);
+  done_ = false;
```
Comment (eggrobin) [inline base/pull_serializer_body.hpp]: lock done

File: base/pull_serializer_body.hpp [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    done_ = true;
+  });
+}
+
+Bytes PullSerializer::Pull() {
+  const Bytes* result;
+  {
+    std::unique_lock<std::mutex> l(lock_);
+    holder_is_full_.wait(l, [this](){ return holder_ != nullptr; });
+    result = holder_.release();
```
Comment (eggrobin) [inline base/pull_serializer_body.hpp]: leakleakleak

File: base/pull_serializer.hpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+ private:
+  // Sets the chunk of data to be returned to |Pull|.  Used as a callback for
+  // the underlying |DelegatingTwoArrayOutputStream|.
+  void Push(Bytes const bytes);
+
+  std::unique_ptr<std::uint8_t[]> data_;
+  internal::DelegatingTwoArrayOutputStream stream_;
+  std::unique_ptr<std::thread> thread_;
+
+  // Synchronization objects for the |holder_|, which contains the |Data|
```
Comment (eggrobin) [inline base/pull_serializer.hpp]: s/Data/Bytes

Comment (pleroy) [inline base/pull_serializer.hpp]: Done.

File: base/pull_serializer_body.hpp (hunk 7 above)
Comment (pleroy) [inline base/pull_serializer_body.hpp]: Moved to the constructor.

File: base/pull_serializer_body.hpp (hunk 8 above)
Comment (pleroy) [inline base/pull_serializer_body.hpp]: Good catch.

**Pull request merged**